### ISSUE TITLE
Graph has unexpected gaps when slightly adjusting time range

### ISSUE THREAD

**Issue by monotok** [VictoriaMetrics/VictoriaMetrics#5428]
### Describe the bug

Victoria returns gaps with certain time windows. The RAW data contains no gaps and I expect the data to be returned without gaps.

### To Reproduce

Pull latest master release of VM.

`make docker-cluster-up`

Use this script to add the data.

```
#!/bin/bash

while IFS="," read -r rec_column1 rec_column2 rec_column3 rec_column4
do
	echo "Posting the metric $rec_column1 with value $rec_column2 and epoch time $rec_column3. Human Readable: $rec_column4"
	curl 'http://localhost:8480/insert/0/prometheus/api/v1/import/prometheus' -d "$rec_column1 $rec_column2 $rec_column3"
done < mydata.csv
```
[mydata.csv](https://github.com/VictoriaMetrics/VictoriaMetrics/files/13559012/mydata.csv)

# Observations

I have take the data and added it to a CSV comparing the queries data with the RAW to help with understanding the issue.
[mydata-shows-gaps-export.csv](https://github.com/VictoriaMetrics/VictoriaMetrics/files/13559562/mydata-shows-gaps-export.csv)

- I have only adjusted the start value by 1 second in each of the examples below.
- The RAW doesn't have any gaps in it.
- Difference between each RAW sample is between 59 and 63 seconds.11

## Correct Graph

Time range:
1701320641 - 1701322722 (5:04:01.641 - 5:38:42.722)

The data seems to be offset by +1 minute.

With curl:

`curl http://localhost:32787/select/0/prometheus/api/v1/query_range -d 'query=mydata[60s]' -d 'start=1701320641' -d 'end=1701322722' -d 'step=60s' -d 'max_lookback=90s' | jq`

With VMUI

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/9199049/f823a754-0139-46f4-b653-9ba36b507838)

## Graph with lots of gaps

Time range:
1701320642 - 1701322722 (5:04:02.642 - 5:38:42.722)

With curl:

`curl http://localhost:32787/select/0/prometheus/api/v1/query_range -d 'query=mydata[60s]' -d 'start=1701320642' -d 'end=1701322722' -d 'step=60s' -d 'max_lookback=90s' | jq`

With VMUI:

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/9199049/1b6c7785-ca38-47e7-b4b0-7622a6211df4)

## Graph is better but still has a few gaps

Time range:
1701320643 - 1701322721 (5:04:03.643 - 5:38:41.721)

With curl:

The end time doesn't seem to make a difference.

`curl http://localhost:32787/select/0/prometheus/api/v1/query_range -d 'query=mydata[60s]' -d 'start=1701320643' -d 'end=1701322721' -d 'step=60s' -d 'max_lookback=90s' | jq`

With VMUI:

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/9199049/ed8593d4-a72f-49a3-b242-d6e877b05699)

### Version

All versions, latest in master.
I have tried, 1.85.2, 1.85.3, 1.95.1

### Logs

No errors.

**Comment by AlxBystrov** [comment]
we have the same problem on version 1.95.1

**Comment by f41gh7** [comment]
Related issue https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5388

 Current workaround - disable cache.

**Comment by bon3o** [comment]
Also, as a workaround, downgrading only vmselect components to v1.94.0-cluster fixes the issue

**Comment by monotok** [comment]
Thanks for everyone's replies. I don't think this is related to #5388 as that seems to be affecting the results when the vmselect component is booted. I have reproduced this issue with versions 1.85.2, 1.85.3 & 1.95.1 while downgrading to 1.90.0 fixes #5388.
I have recreated the vmselect containers with 1.90.0 and disabled the cache within VMUI; this doesn't fix the behavior I am seeing.

**Comment by hagen1778** [comment]
@monotok could you verify the resolution of your data by running the following query `scrape_interval(mydata[1m])`?
Btw, using vectors like `mydata[60s]` for query_range API doesn't make much sense. Is there anything specific you want to achieve with this query?

**Comment by monotok** [comment]
@hagen1778 I wasn't sure if you wanted me to provide the scrape interval with a set step or not so I have provided a few variations. In summary the data consists of "min:58, median:60, max:63".

```
curl http://localhost:32798/select/0/prometheus/api/v1/query_range -d 'query=scrape_interval(mydata[60s])'  -d 'start=1701320641' -d 'end=1701322722' | jq
```

```json
{
  "status": "success",
  "isPartial": false,
  "data": {
    "resultType": "matrix",
    "result": [
      {
        "metric": {},
        "values": [
          [
            1701320941,
            "61"
          ],
          [
            1701321241,
            "59"
          ],
          [
            1701321541,
            "60"
          ],
          [
            1701321841,
            "60"
          ],
          [
            1701322141,
            "59"
          ],
          [
            1701322441,
            "59"
          ]
        ]
      }
    ]
  },
  "stats": {
    "seriesFetched": "1",
    "executionTimeMsec": 3
  }
}
```

```
curl http://localhost:32798/select/0/prometheus/api/v1/query_range -d 'query=scrape_interval(mydata[60s])'  -d 'start=1701320641' -d 'end=1701322722' -d 'step=60s' | jq
```

```json
{
  "status": "success",
  "isPartial": false,
  "data": {
    "resultType": "matrix",
    "result": [
      {
        "metric": {},
        "values": [
          [
            1701320821,
            "60"
          ],
          [
            1701320881,
            "59"
          ],
          [
            1701320941,
            "61"
          ],
          [
            1701321001,
            "59"
          ],
          [
            1701321061,
            "61"
          ],
          [
            1701321121,
            "60"
          ],
          [
            1701321181,
            "60"
          ],
          [
            1701321241,
            "59"
          ],
          [
            1701321301,
            "60"
          ],
          [
            1701321361,
            "62"
          ],
          [
            1701321421,
            "59"
          ],
          [
            1701321481,
            "60"
          ],
          [
            1701321541,
            "60"
          ],
          [
            1701321601,
            "59"
          ],
          [
            1701321661,
            "61"
          ],
          [
            1701321721,
            "59"
          ],
          [
            1701321781,
            "60"
          ],
          [
            1701321841,
            "60"
          ],
          [
            1701321901,
            "61"
          ],
          [
            1701321961,
            "59"
          ],
          [
            1701322021,
            "60"
          ],
          [
            1701322081,
            "61"
          ],
          [
            1701322141,
            "59"
          ],
          [
            1701322201,
            "60"
          ],
          [
            1701322261,
            "63"
          ],
          [
            1701322321,
            "58"
          ],
          [
            1701322381,
            "60"
          ],
          [
            1701322441,
            "59"
          ],
          [
            1701322501,
            "61"
          ],
          [
            1701322561,
            "60"
          ],
          [
            1701322621,
            "59"
          ],
          [
            1701322681,
            "61"
          ]
        ]
      }
    ]
  },
  "stats": {
    "seriesFetched": "1",
    "executionTimeMsec": 0
  }
}
```

We have 1min interval data that we push into VM and therefore we set the step to be 60s. I understand that VM calculates the step based on the first 20 data points? Although I have read here [keyconcepts range-query](https://new.docs.victoriametrics.com/keyconcepts/#range-query) it defaults to 5 mins. We set to 60s for consistency of the graphs in case we have missed data or longer than usual intervals in the data. 

Could you elaborate on how using vectors doesn't make much sense?

**Comment by hagen1778** [comment]
Hi @monotok! Thanks for provided data!

> We have 1min interval data that we push into VM and therefore we set the step to be 60s. I understand that VM calculates the step based on the first 20 data points? Although I have read here [keyconcepts range-query](https://new.docs.victoriametrics.com/keyconcepts/#range-query) it defaults to 5 mins. We set to 60s for consistency of the graphs in case we have missed data or longer than usual intervals in the data.

The `5min` default is for instant queries. For range queries it looks on the median resolution between samples based on first 20 datapoints. And this resolution, based on the data you provided, is about 67sec. So VM will look for a datapoint up to 67sec in past by default.

But when you specify lookback window explicilty `mydata[60s]`, it overrides this behavior. So it starts to look for 60s back, and misses those datapoints which are older than 60s. Hence, the gaps. If you remove [60s] from your query - it should work without gaps.

> Btw, using vectors like mydata[60s] for query_range API doesn't make much sense. Is there anything specific you want to achieve with this query?

Back to this - any reasons to specify the lookback window like this? Please note, the window [duration] need to be specified only for rollup queries - see https://docs.victoriametrics.com/MetricsQL.html#rollup-functions

**Comment by monotok** [comment]
Thanks for the quick response.

> The `5min` default is for instant queries. For range queries it looks on the median resolution between samples based on first 20 datapoints. And this resolution, based on the data you provided, is about 67sec. So VM will look for a datapoint up to 67sec in past by default.

Could you explain how you have got to the figure of 67 seconds? If I am understanding this correctly, a median is calculated by sorting the data and then picking the middle from the list. The highest interval I have is 63 seconds so I can't see how it can be 67 seconds.

> But when you specify lookback window explicilty `mydata[60s]`, it overrides this behavior. So it starts to look for 60s back, and misses those datapoints which are older than 60s. Hence, the gaps. If you remove [60s] from your query - it should work without gaps.

The raw data has no missing data points so this shouldn't matter?

> > Btw, using vectors like mydata[60s] for query_range API doesn't make much sense. Is there anything specific you want to achieve with this query?

We have decided we only want to look back as far as 1 minute and are happy to see real gaps.

**Comment by hagen1778** [comment]
> Could you explain how you have got to the figure of 67 seconds? If I am understanding this correctly, a median is calculated by sorting the data and then picking the middle from the list. The highest interval I have is 63 seconds so I can't see how it can be 67 seconds.

It gets additionally adjusted to exclude possible jitter during scraping - see below 
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/1bed3a1da720ea141f52dc9ce6736768ccbd6d6b/app/vmselect/promql/rollup.go#L654-L655
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/1bed3a1da720ea141f52dc9ce6736768ccbd6d6b/app/vmselect/promql/rollup.go#L819-L839

After `getMaxPrevInterval` is called, lookback is bumped by 7.5sec, which is 1/8 of median of scrape interval.

> The raw data has no missing data points so this shouldn't matter?

The raw data has inconsistent range between samples. The range could be 60s, 61s, 62s, 63s. When you send query `mydata` it will look back up to 67.5 sec for previous datapoint. When you send `mydata[60s]` - it looks back up to 60s and will miss some of those data points.

**Comment by monotok** [comment]
@hagen1778 Okay, that makes sense, thanks for clearing that up. However I don't think that explains the odd behaviour that I initially reported.

All these examples are using the exact same RAW data as shared above.

The below graphs has lots of gaps in it when using the lookback of 60s and removing the lookback window fills in those gaps.

Time window for this query: 1701320642 - 1701322722 (5:04:02.642 - 5:38:42.722)

RED query is using the default lookback and has no gaps. 
BLUE query has lots of gaps which you can see as it is overlay-ed on the top. (Same as second image in the opening issue).

![Screenshot 2024-02-07 at 17 27 03](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/9199049/cb4a3bdb-4d80-4a0e-a57b-b17ced16d432)

So from your explanation I can see why the gaps get filled in the way they do. 

However, changing the time window of the query by 1 second, substantially changes the behaviour.

Time window for this query starts 1 second before: 1701320641 - 1701322722 (5:04:01.641 - 5:38:42.722)

What I am sometimes seeing day to day is mostly the correct graphs and then briefly it goes gappy as shown and back to normal, I managed it replicate the behaviour with this data.

This is the graph I would expect as it shows the correct data; however each metric is offset by +1min from the RAW data. 

From what you have said there probably should be gaps.

So the question is, why is the time window; specifically the `start=` having such an impact?

![Screenshot 2024-02-07 at 17 36 28](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/9199049/c590dd99-c5ee-4eec-8ee4-948a56777d63)

Thank you for your time and hopefully that makes sense!

**Comment by hagen1778** [comment]
> So the question is, why is the time window; specifically the start= having such an impact?

See description https://docs.victoriametrics.com/keyconcepts/#range-query from below:
> The result of Range query is a list of [time series](https://docs.victoriametrics.com/keyconcepts/#time-series) matching the filter in query expression. Each returned series contains (timestamp, value) results for the query executed at start, start+step, start+2*step, …, end timestamps. In other words, Range query is an [Instant query](https://docs.victoriametrics.com/keyconcepts/#instant-query) executed independently at start, start+step, …, end timestamps.

If lookback window is limited by 60s and raw datapoints sometimes exceed 60s interval between them, so depending on **when** you look for a datapoint it could either find it or not. I.e. if you have datapoints at 11:59:59 and 12:01:00, then requests at &time=12:02:00 will find data point at 12:01:00 (it is on the 60s time range). And request at &time=12:01:00 - will not find datapoint, because it is 61s away.

**Comment by monotok** [comment]
@hagen1778 Apologies for the delayed response.

Thanks for explaining the concepts. I can now see what is going on with the example I shared.

Will post what I found below, it might help someone else.

Grafana is not aligning the start time to the minute so it can be any value, in my example it is `1701320642`.

Eg 11/30/2023 5:04:02.642

Then the range query is `1701320642`, `1701320642+60`, `1701320642+(2x60)` etc.

They match the RAW data until one of the RAW data points is 11/30/2023 5:06:03.763. It is ahead by 1 second and therefore a gap appears.

The graph that `worked` was offset by +1 minute so all the data points were ephemeral.  The RAW data points were at 2 seconds past the minute while the start + step was at 1 second past the minute.
